unit tests for the annotation related functions in pkg/util

vjayaramrh opened the pull request:
Signed-off-by: Vishwanath Jayaraman <vjayaram@redhat.com>

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/ovn-org/ovn-kubernetes/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Trivial changes are exempt from following this template.
If your change is non-trivial, please provide the following information:
-->

**- What this PR does and why is it needed**
<!--
A summary of the changes within this pull request and some context
as to why they were made
-->
The PR adds unit tests (code refactor is not the scope for this PR) for most of the functions defined in the following files:

    - pkg/util/node_annotations.go 
    - pkg/util/pod_annotation.go
    -pkg/util/subnet_annotations.go


**- Special notes for reviewers**
<!--
What exactly did you change - you may also defer to information
contained in commit messages. At a bare minimum it's worth highlighting
which areas of the code were changed as it's easier to assign reviewers
-->


**- How to verify it**
<!--
Did you include unit tests? or end-to-end tests?
How can I manually verify that this patch achieves its objective
-->
Pull this PR into your workspace
Navigate to go-controller/pkg/util directory
execute below commands
go test -coverprofile cp.out
go tool cover -html=cp.out
select files node_annotations.go, pod_annotation.go, subnet_annotations.go from drop down to verify coverage

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

vjayaramrh commented:
@trozet @danwinship would appreciate your review

danwinship commented on go-controller/pkg/util/node_annotations_unit_test.go:
```diff
@@ -0,0 +1,298 @@
+	"reflect"
+	"testing"
+)
+import "fmt"
```
the imports here are all messed up

vjayaramrh replied:
shall fix the imports

danwinship commented on go-controller/pkg/util/node_annotations_unit_test.go:
```diff
@@ -0,0 +1,298 @@
+				Mode:   config.GatewayModeLocal,
+				VLANID: &vlanid,
+			},
+			expOutput: []byte(`{"mode":"local","node-port-enable":"false","vlan-id":"1024"}`),
```
This should actually be an error: `VLANID` is only allowed with shared gateway mode

vjayaramrh replied:
Thanks for catching this, however, I should be able to modify line 39 to be `Mode: config.GatewayModeLocal` and L42 to be `expOutput: []byte(`{"mode":"local","node-port-enable":"false","vlan-id":"1024"}`)` , right?

danwinship replied:
Assuming that when you say `GatewayModeLocal` you mean `GatewayModeShared` and that when you say `"mode":"local"` you mean `"mode":"shared'`, then yes.

But also, I meant that this is pointing out a bug in the (non-test) code; it ought to be throwing an error in the case you have now.

vjayaramrh replied:
Thanks for the correction again, Below should be the corrected test case
```
desc: "test VLANID not nil",
			inpL3GwCfg: &L3GatewayConfig{
				Mode:   config.GatewayModeShared,
				VLANID: &vlanid,
			},
			expOutput: []byte(`{"mode":"shared","node-port-enable":"false","vlan-id":"1024"}`),
```
Let me open a Github issue to track that the non-test code needs to throw an error for the below scenario
```
inpL3GwCfg: &L3GatewayConfig{
				Mode:   config.GatewayModeLocal,
				VLANID: &vlanid,
			},
```

vjayaramrh replied:
FYI Opened issue https://github.com/ovn-org/ovn-kubernetes/issues/1478

danwinship commented on go-controller/pkg/util/node_annotations_unit_test.go:
```diff
@@ -0,0 +1,298 @@
+			inputParam:  []byte(`{"mode":"local","mac-address":"11:22:33:44:55:66","ip-addresses":["192.168.1/24","fd01::1234/64"]}`),
+		},
+		{
+			desc:        "test bad 'next-hop value",
```
(missing close single-quote)

vjayaramrh replied:
Shall fix this

danwinship commented on go-controller/pkg/util/node_annotations_unit_test.go:
```diff
@@ -0,0 +1,298 @@
+				ChassisID: " ",
+			},
+		},
+	}
```
This is clearly not a complete set of tests for `SetL3GatewayConfig`...

vjayaramrh replied:
Let me relook into the function code and see what I may have missed. Seems like I only have success cases currently.

vjayaramrh replied:
![Screen Shot 2020-07-02 at 4 41 34 PM](https://user-images.githubusercontent.com/62150532/86411659-f5925600-bc82-11ea-9de9-9eafc1adb330.png)

From the coverage report, I am seeing that I am missing error path testing. I am guessing that's probably what you maybe referring to. I shall create the mocks for the Annotator interface in pkg/kube to test the error path

danwinship replied:
well, I was thinking of `SetL3GatewayConfig` as being the function that did the marshalling, but I guess you effectively test that via the `MarshalJSON` test

vjayaramrh replied:
> well, I was thinking of `SetL3GatewayConfig` as being the function that did the marshalling, but I guess you effectively test that via the `MarshalJSON` test

Yes, MarshalJSON() code path is 100% covered as part of the unit test
![Screen Shot 2020-07-07 at 10 09 17 AM](https://user-images.githubusercontent.com/62150532/86801977-2d5e1c80-c03a-11ea-9efe-7e6b18be0f3c.png)

vjayaramrh replied:
Additionally, I have added tests to cover the error paths for the `SetL3GatewayConfig` function
![Screen Shot 2020-07-07 at 10 14 32 AM](https://user-images.githubusercontent.com/62150532/86802637-d4db4f00-c03a-11ea-918e-7020217ac679.png)

danwinship commented on go-controller/pkg/util/node_annotations_unit_test.go:
```diff
@@ -0,0 +1,298 @@
+		{
+			desc:        "error: test unsupported gateway mode",
+			inputParam:  []byte(`{"mode":"blah"}`),
+			errExpected: true,
```
It's generally better to assert more specific things about the error.

eg, if someone added a check that "mac-address" always has to be set, and then later broke the check for unsupported gateway mode, this test would still pass, because it would be returning a "mac address is unset" error, which is the wrong error, but would fool the test case.

in other places it's common to check for a substring match on the returned error

vjayaramrh replied:
Valid point. If an error message is originating inside of the function being unit tested, checking for substring match would be effective.

danwinship commented on go-controller/pkg/util/subnet_annotations_unit_test.go:
```diff
@@ -0,0 +1,237 @@
+
+func TestSetNodeLocalNatAnnotation(t *testing.T) {
+
+}
```
?

vjayaramrh replied:
Thanks for the catch, shall fix it.

vjayaramrh commented:
> It still does not make sense to have two separate sets of unit tests. Either these tests should be merged into the existing ones, or the existing ones should be deleted and these ones extended to cover all of the things that the old ones tested as well

@trozet had made a comment regarding the approach on an earlier PR at https://github.com/ovn-org/ovn-kubernetes/pull/1382#issuecomment-648384112

danwinship commented:
> @trozet had made a comment regarding the approach on an earlier PR

> current approach by @vjayaramrh is to add unit tests via testify, and remove the leftover duplicates in ginkgo later. That way it will be more obvious if he removes duplicate tests in ginkgo in a separate PR, we can see coveralls change 0% and its easy merge.

So the new tests have 100% coverage on their own? They seem to be missing things that the old tests have. eg, the "`Nil entry in GW`" test in `pod_annotation_test.go`.

Even if the new tests have 100% line coverage, they may be missing specific code flows that were tested by the old tests...

vjayaramrh commented:
> > @trozet had made a comment regarding the approach on an earlier PR
> 
> > current approach by @vjayaramrh is to add unit tests via testify, and remove the leftover duplicates in ginkgo later. That way it will be more obvious if he removes duplicate tests in ginkgo in a separate PR, we can see coveralls change 0% and its easy merge.
> 
> So the new tests have 100% coverage on their own? They seem to be missing things that the old tests have. eg, the "`Nil entry in GW`" test in `pod_annotation_test.go`.
> 
> Even if the new tests have 100% line coverage, they may be missing specific code flows that were tested by the old tests...

Below is the coverage from the newly written unit tests for node_annotations.go, pod_annotations.go and subnet_annotations.go
![Screen Shot 2020-07-02 at 2 02 23 PM](https://user-images.githubusercontent.com/62150532/86399442-e3a5b880-bc6c-11ea-8b77-e5b5423ddcac.png)
I have tried to cover as much code as possible for the functions in those files. The coverage can easily be seen as HTML output by executing the below commands (rename the pod_annotation_test.go, node_annotations_test.go, subnet_annotations_test.go as pod_annotation_test.go.BAK, node_annotations_test.go.BAK, subnet_annotations_test.go.BAK  ) in the go-controller/pkg/util folder
```
go test -coverprofile cp.out
go tool cover -html=cp.out
```
Where feasible and with no code refactor, the unit tests try to cover as much code path as possible for the function being unit tested.

vjayaramrh commented:
@danwinship I believe that I have addressed all your comments, please let me if there is any other issue that I need to take care that is blocking merge of this PR.

vjayaramrh commented:
/retest

trozet commented on go-controller/pkg/util/node_annotations_unit_test.go:
```diff
@@ -0,0 +1,346 @@
+			inputParam: []byte(`{"mode":"local","vlan-id":"A"}`),
+			errMatch:   fmt.Errorf("bad 'vlan-id' value"),
+		},
+
```
spurious newline

vjayaramrh replied:
shall fix

trozet commented on go-controller/pkg/util/node_annotations_unit_test.go:
```diff
@@ -0,0 +1,346 @@
+			},
+		},
+		{
+			desc:             "success: apply empty Chassis id ",
```
extra trailing white space after Chassis id

vjayaramrh replied:
shall fix

trozet commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -0,0 +1,266 @@
+		expectedOutput map[string]string
+	}{
+		{
+			desc:           "PodAnnotation instance with no fields set",
```
you aligned the fields and values here, but in the subsequent entries they are not aligned

vjayaramrh replied:
gofmt when using the IDE automatically takes care of aligning them appropriately.
When I try to manually re-align them as can be seen below, I get an error when I execute `make gofmt`. This is because on Line31, a struct is being populated on mutli line

![Screen Shot 2020-07-13 at 5 10 59 PM](https://user-images.githubusercontent.com/62150532/87358763-146ad500-c52c-11ea-87e1-a568f87d85a3.png)

trozet commented on go-controller/pkg/util/pod_annotation_unit_test.go:
```diff
@@ -0,0 +1,266 @@
+			},
+		},
+		{
+			desc: "next hop not set for route ",
```
extra space after route?

vjayaramrh replied:
shall fix
